### ISSUE TITLE
Slow mysql startup time causes error message on grain start

### ISSUE THREAD

**Issue by JamborJan** [zarvox/vagrant-spk#1]
When using vagrant-spk as described in the README the following will be the result after launching the first instance of the test app:

```
...** SANDSTORM SUPERVISOR: Starting up grain.
Installing MySQL system tables...
150416  8:21:48 [Warning] Using unique option prefix key_buffer instead of key_buffer_size is deprecated and will be removed in a future release. Please use the full name instead.
OK
Filling help tables...
150416  8:21:49 [Warning] Using unique option prefix key_buffer instead of key_buffer_size is deprecated and will be removed in a future release. Please use the full name instead.
OK

To start mysqld at boot time you have to copy
support-files/mysql.server to the right place for your system

PLEASE REMEMBER TO SET A PASSWORD FOR THE MySQL root USER !
To do so, start the server, then issue the following commands:

/usr/bin/mysqladmin -u root password 'new-password'
/usr/bin/mysqladmin -u root -h sandbox password 'new-password'

Alternatively you can run:
/usr/bin/mysql_secure_installation

which will also give you the option of removing the test
databases and anonymous user created by default.  This is
strongly recommended for production servers.

See the manual for more instructions.

You can start the MySQL daemon with:
cd /usr ; /usr/bin/mysqld_safe &

You can test the MySQL daemon with mysql-test-run.pl
cd /usr/mysql-test ; perl mysql-test-run.pl

Please report any problems at http://bugs.mysql.com/

150416  8:21:49 [Warning] Using unique option prefix key_buffer instead of key_buffer_size is deprecated and will be removed in a future release. Please use the full name instead.
```

I will fork the repo and try to make it self configurable if possible.

**Comment by JamborJan** [comment]
The problem is startup time. After waiting 5 seconds it is working. Basically you need to start the grain, see a bad gateway message, reload it and it's working.

Maybe it's a good idea to check if the mysql setup in the package can be optimized e.g. by applying tweaks @jparyani already found while optimizing mysql for his app porting process via docker. He used to set it up like follows and I also implemented it for an app and it worked pretty well. Also the initial grain size was lower (23MB instead of 30 MB now with vagrant-spk).

```
# Install Mysql
RUN apt-get -y install mysql-server

# Install php
RUN apt-get -y install php5 php5-mysql

# Install nginx
RUN apt-get -y install php5-fpm nginx

# minimize mysql allocations
RUN echo '[mysqld]\ninnodb_data_file_path = ibdata1:10M:autoextend\ninnodb_log_file_size = 10KB\ninnodb_file_per_table = 1' > /etc/mysql/conf.d/small.cnf
RUN sed -i 's_^socket\s*=.*_socket = /tmp/mysqld.sock_g' /etc/mysql/*.cnf && ln -s /tmp/mysqld.sock /var/run/mysqld/mysqld.sock
RUN rm -rf /var/lib/mysql/* && mysql_install_db && chown -R mysql: /var/lib/mysql

# Setup mysql//mysql user
RUN /usr/sbin/mysqld & \
    sleep 10s &&\
    echo "GRANT ALL ON *.* TO mysql@'%' IDENTIFIED BY 'mysql' WITH GRANT OPTION; FLUSH PRIVILEGES; CREATE SCHEMA app;" | mysql
```
(from https://github.com/jparyani/Tiny-Tiny-RSS/blob/sandstorm/Dockerfile)

I will work on that and let you know the result.

**Comment by JamborJan** [comment]
I applied the hack but I cannot see any big change.

Furthermore is the described issue only present when running ```vagrant-spk dev```. When the app is packed it is running smoothly.

So I guess this was a bunch of hot air and the issue can be closed without any further steps. Anyway I have send a PR. You can decide yourself whether to take it or discard it.

**Comment by JamborJan** [comment]
Sorry for confusion. It's getting weird for me. Somehow mysql is nit starting at all.

When starting the app (no matter if DEV or in a package) I see that the mysql demo calculation 1+1 is returning an error.

When I ssh into the vagrant machine and try to access mysql it's not running and it can also not be started. ```dmesg``` shows:

```
[  543.299687] init: mysql main process (7821) terminated with status 1
[  543.299716] init: mysql main process ended, respawning
[  543.307192] init: mysql post-start process (7822) terminated with status 1
[  543.320993] type=1400 audit(1429179746.426:17): apparmor="STATUS" operation="profile_replace" profile="unconfined" name="/usr/sbin/mysqld" pid=7837 comm="apparmor_parser"
[  543.347234] init: mysql main process (7849) terminated with status 1
[  543.347253] init: mysql main process ended, respawning
[  543.353097] init: mysql post-start process (7850) terminated with status 1
[  543.364855] type=1400 audit(1429179746.470:18): apparmor="STATUS" operation="profile_replace" profile="unconfined" name="/usr/sbin/mysqld" pid=7865 comm="apparmor_parser"
[  543.383385] init: mysql main process (7877) terminated with status 1
[  543.383448] init: mysql respawning too fast, stopped
[  634.052310] type=1400 audit(1429179837.158:19): apparmor="STATUS" operation="profile_replace" profile="unconfined" name="/usr/sbin/mysqld" pid=7923 comm="apparmor_parser"
[  634.076569] init: mysql main process (7935) terminated with status 1
[  634.076589] init: mysql main process ended, respawning
[  634.083636] init: mysql post-start process (7936) terminated with status 1
[  634.097558] type=1400 audit(1429179837.202:20): apparmor="STATUS" operation="profile_replace" profile="unconfined" name="/usr/sbin/mysqld" pid=7951 comm="apparmor_parser"
[  634.123165] init: mysql main process (7963) terminated with status 1
[  634.123183] init: mysql main process ended, respawning
[  634.131160] init: mysql post-start process (7964) terminated with status 1
[  634.143563] type=1400 audit(1429179837.246:21): apparmor="STATUS" operation="profile_replace" profile="unconfined" name="/usr/sbin/mysqld" pid=7979 comm="apparmor_parser"
[  634.161650] init: mysql main process (7991) terminated with status 1
[  634.161669] init: mysql respawning too fast, stopped
```

Any ideas from your side? I have to postpone working on that issue as I had other plans today (port of ownCloud with sqlite and not mysql).

JJ

**Comment by JamborJan** [comment]
Okay. My fault. It's working now after having destroyed everything and set it up from scratch.

Key take away: when you are running ```vagrant-spk dev```you can get a ```502 Bad Gateway```error. reload the page and you are fine.

**Comment by paulproteus** [comment]
I'm not totally sure it's your fault. Either way, we should make sure this is smoother. Thanks a great deal for testing out this preliminary version!

**Comment by zarvox** [comment]
Hey, thanks for testing this all out!  I'm glad to hear that you got it slightly working, despite the lack of polish. :)

You've rightly discovered that the launcher.sh script has a race condition in which nginx is up and accepts requests before php5-fpm and/or mysql are necessarily up and listening on their sockets, and I can confirm that I also sometimes hit 502's on the first pageload after launching the grain (and this is true in both dev mode and with the packed app).  It should certainly be fixed before using vagrant-spk to publish an official app package.

I can think of doing some bash hackery to wait for those sockets to be created before starting nginx (which is the signal to sandstorm that the grain is ready).  Another approach would be to run something like systemd in the container, which will create all the sockets in advance and hand them off to the processes started, so there's no gap between "nginx listening on socket" and "requests will reliably be delivered", as there is with the current launcher script.

Regarding the large grain size: I'm happy to merge pull requests for getting the default mysql table size down; this is a great idea!

**Comment by JamborJan** [comment]
I will investigate more during working on the port. Status: 
https://github.com/JamborJan/vagrant-spk
https://github.com/JamborJan/owncloud-sandstorm

Nothing is working so far. When you build the app it actually wants to download the index php and the database is not up and running, I think I screwed up the vagrant-spk fork. I will check this in more detail later.

Anyway: this issue is closed. More about the mysql optimization will be mentioned in the two repos when I find something.
